src/dbnode/persist/fs/files.go

```diff
@@ -1375,12 +1476,48 @@ func filesetFileForTime(t time.Time, suffix string) string {
 	return fmt.Sprintf("%s%s%d%s%s%s", filesetFilePrefix, separator, t.UnixNano(), separator, suffix, fileSuffix)
 }
 
-func filesetPathFromTime(prefix string, t time.Time, suffix string) string {
+func filesetPathFromTimeLegacy(prefix string, t time.Time, suffix string) string {
 	return path.Join(prefix, filesetFileForTime(t, suffix))
 }
 
 func filesetPathFromTimeAndIndex(prefix string, t time.Time, index int, suffix string) string {
-	return path.Join(prefix, filesetFileForTime(t, fmt.Sprintf("%d%s%s", index, separator, suffix)))
+	newSuffix := fmt.Sprintf("%d%s%s", index, separator, suffix)
+	return path.Join(prefix, filesetFileForTime(t, newSuffix))
+}
+
+// isFirstVolumeLegacy returns whether the first volume of the provided type is
+// legacy, i.e. does not have a volume index in its filename. It only checks for
+// the existence of the file and the non-legacy version of the file, and returns
+// and error if neither exist.
+func isFirstVolumeLegacy(prefix string, t time.Time, suffix string) (bool, error) {
+	// Check non-legacy path first to optimize for newer files.
+	path := filesetPathFromTimeAndIndex(prefix, t, 0, suffix)
+	_, err := os.Stat(path)
+	if err == nil {
+		return false, nil
+	}
+
+	legacyPath := filesetPathFromTimeLegacy(prefix, t, suffix)
+	_, err = os.Stat(legacyPath)
+	if err == nil {
+		return true, nil
```

Comment:
Do you also need to check/verify for a completed checkpoint file?

Reply:
This is a good point. This function should be used solely to figure out whether the first volume is legacy or not before we actually use them, so checking for a complete checkpoint file should come after this function is called. I'll add a comment.

Reply:
I assume its fine because this is only used by the reader to determine what code path to follow, and if the checkpoint file is not complete the reader will figure that out pretty quickly and error out?